compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala

```diff
@@ -0,0 +1,358 @@
+package dotty.tools.dotc
+package transform
+
+import ast.tpd
+import ast.Trees.*
+import core.*
+import Flags.*
+import Decorators.*
+import Contexts.*
+import Symbols.*
+import Constants.Constant
+import Decorators.*
+import DenotTransformers.IdentityDenotTransformer
+import Names.*
+import dotty.tools.dotc.core.NameKinds.DefaultGetterName
+
+import dotty.tools.dotc.core.Types.{MethodType, NamedType, PolyType, Type, NoPrefix, NoType}
+
+import dotty.tools.dotc.printing.Formatting.hl
+
+import scala.collection.mutable
+import scala.util.boundary, boundary.break
+import dotty.tools.dotc.core.StdNames.nme
+import dotty.tools.unreachable
+
+/**Implementation of SIP-61.
+ * Runs when `@unroll` annotations are found in a compilation unit, installing new definitions
+ *
+ * Note that it only generates `Invisible` methods, so no interactions with Zinc/SemanticDB
+ */
+class UnrollDefinitions extends MacroTransform, IdentityDenotTransformer {
+  self =>
+
+  import tpd.*
+
+  private var _unrolledDefs: util.EqHashMap[Symbol, ComputedIndicies] | Null = null
+  private def initializeUnrolledDefs(): util.EqHashMap[Symbol, ComputedIndicies] =
+    val local = _unrolledDefs
+    if local == null then
+      val map = new util.EqHashMap[Symbol, ComputedIndicies]
+      _unrolledDefs = map
+      map
+    else
+      local.clear()
+      local
+
+  override def phaseName: String = UnrollDefinitions.name
+
+  override def description: String = UnrollDefinitions.description
+
+  override def changesMembers: Boolean = true
+
+  override def run(using Context): Unit =
+    if ctx.compilationUnit.hasUnrollDefs then
+      super.run // create and run the transformer on the current compilation unit
+
+  def newTransformer(using Context): Transformer =
+    UnrollingTransformer(ctx.compilationUnit.unrolledClasses.nn)
+
+  type ComputedIndicies = List[(Int, List[Int])]
+  type ComputeIndicies = Context ?=> Symbol => ComputedIndicies
+
+  private class UnrollingTransformer(classes: Set[Symbol]) extends Transformer {
```

Comment:
Consider giving a better name to the `classes` parameter. Here it's decent but at use site it's confusing what it's supposed to be.